# NAMA          : KEVIN SYNAGOGUE PANJAITAN
# NIM           : F1C221058
# PROGRAM STUDI : MATEMATIKA

## pemangilan beberapa paket yang saya gunakan 
Pandas (import pandas as pd) digunakan untuk manipulasi dan analisis data tabular, memungkinkan pengguna untuk membaca, membersihkan, dan memanipulasi data dengan mudah. Scikit-learn menyediakan berbagai alat untuk pembelajaran mesin, termasuk train_test_split untuk membagi data menjadi subset pelatihan dan pengujian. LogisticRegression dari scikit-learn digunakan untuk model klasifikasi biner atau multikelas, memprediksi probabilitas kelas berdasarkan variabel input. Setelah melatih model, accuracy_score digunakan untuk menghitung akurasi model, sedangkan confusion_matrix memberikan matriks kebingungannya untuk mengevaluasi prediksi dengan membandingkan hasil yang diprediksi dengan nilai sebenarnya. Terakhir, matplotlib.pyplot digunakan untuk visualisasi data atau hasil, dan math digunakan untuk perhitungan matematis dasar.

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix
import matplotlib.pyplot as plt
import math


# pembaca data file csv

In [2]:
df = pd.read_csv('rancangan dummy untuk tugas reg logistik.xlsx - data.csv')

In [3]:
# baca data yang di berikan 
df1 = pd.read_csv('keterangan.csv')

## kita melihat data variabel apa yang ada pada data dan kita pilah pilih apa yang akan di analisis lebih lanjut 

In [4]:
# meilhat cariabel data yang di berikan 
df.columns

Index(['X1_klas', 'X2_JK_KRT', 'X3_kelum_KRT', 'X4_ijz_KRT', 'X5_bekerja_KRT',
       'X6_inet_KRT', 'X7_milik_jamkes_KRT', 'X8_kepemilikan_rumah',
       'X9_luas_kapita_cat', 'X10_atap', 'X11_dinding', 'X12_lantai',
       'X13_sanitasi', 'X14_air', 'X15_bbm', 'X16_layanan_keuangan',
       'X17_motor', 'X18_tanah', 'X19_kks', 'X20_pkh', 'X21_bpnt',
       'X22_bantuan_pemda', 'X23_kelompok_pengeluaran'],
      dtype='object')

In [5]:
# melihat seluruh nilai pada data 
df

X1_klas  X2_JK_KRT  X3_kelum_KRT  X4_ijz_KRT  X5_bekerja_KRT  \
0           2          1             2           1               1   
1           2          1             2           1               2   
2           2          1             2           1               1   
3           1          1             2           1               1   
4           1          2             2           2               2   
...       ...        ...           ...         ...             ...   
3264        2          1             2           1               1   
3265        2          1             2           2               1   
3266        2          1             2           2               1   
3267        1          1             2           2               1   
3268        2          1             2           2               1   

      X6_inet_KRT  X7_milik_jamkes_KRT  X8_kepemilikan_rumah  \
0               1                    1                     1   
1               2                    2                     1   
2               1                    1                     1   
3               2                    1                     2   
4               2                    1                     1   
...           ...                  ...                   ...   
3264            2                    1                     1   
3265            2                    2                     1   
3266            2                    1                     1   
3267            2                    2                     2   
3268            2                    2                     1   

      X9_luas_kapita_cat  X10_atap  ...  X14_air  X15_bbm  \
0                      2         1  ...        2        1   
1                      2         1  ...        2        1   
2                      2         1  ...        1        1   
3                      2         1  ...        1        2   
4                      2         1  ...        1        1   
...                  ...       ...  ...      ...      ...   
3264                   2         1  ...        1        1   
3265                   2         1  ...        1        1   
3266                   2         1  ...        1        1   
3267                   1         1  ...        2        1   
3268                   2         1  ...        1        1   

      X16_layanan_keuangan  X17_motor  X18_tanah  X19_kks  X20_pkh  X21_bpnt  \
0                        1          1          1        2        2         2   
1                        2          2          1        2        2         2   
2                        1          1          2        2        2         2   
3                        2          1          2        2        2         2   
4                        2          1          1        2        2         2   
...                    ...        ...        ...      ...      ...       ...   
3264                     2          1          1        2        2         2   
3265                     2          1          1        2        2         2   
3266                     2          1          1        2        2         2   
3267                     2          1          2        2        2         2   
3268                     1          2          2        2        2         1   

      X22_bantuan_pemda  X23_kelompok_pengeluaran  
0                     2                         2  
1                     2                         1  
2                     2                         2  
3                     2                         2  
4                     2                         2  
...                 ...                       ...  
3264                  2                         2  
3265                  2                         2  
3266                  2                         1  
3267                  2                         1  
3268                  2                         1  

[3269 rows x 23 columns]

## keterangan pada tabel diatas yaitu 

In [6]:
# mengahpus kolom nomor dan untuk memudahkan pembacaaan keterangan 
df1 = df1.drop('No', axis=1)
df1

Nama Variabel                        Kategori   \
0                                  NaN                                1   
1                       Tempat tinggal                             Kota   
2                    jenis kelamin KRT                             Laki   
3                             Umur KRT                         <= 30 th   
4                       pendidikan KRT                      SD ke bawah   
5                        Pekerjaan KRT                         bekerja    
6                   akses internet KRT                            akses   
7                           jamkes KRT                           punya    
8                         status rumah                    milik sendiri   
9                    luas lantai rumah                <10 m2 per kapita   
10                          atap rumah                            layak   
11                       dinding rumah                            layak   
12                        lantai rumah                            layak   
13                      sanitasi rumah                            layak   
14                          air minum                             layak   
15                 bahan bakar memasak                  listrik dan gas   
16              akses layanan keuangan                            akses   
17                         punya motor                               ya   
18                   punya tanah/lahan                               ya   
19                          punya KKS                               ya    
20                        menerima PKH                               ya   
21                       menerima BPNT                               ya   
22  menerima bantuan pemerintah daerah                               ya   
23                    pengeluaran ruta  40% terendah (kurang sejahtera)   

                                      Unnamed: 3  \
0                                              2   
1                                           Desa   
2                                      perempuan   
3                                        > 30 th   
4                                    SMP ke atas   
5                                  tidak bekerja   
6                                    tidak akses   
7                                    tidak punya   
8                           selain milik sendiri   
9                             >= 10m2 per kapita   
10                                   tidak layak   
11                                   tidak layak   
12                                   tidak layak   
13                                   tidak layak   
14                                   tidak layak   
15                        selain listrik dan gas   
16                                         tidak   
17                                         tidak   
18                                         tidak   
19                                         tidak   
20                                         tidak   
21                                         tidak   
22                                         tidak   
23  40%menengah+20%teratas (cukup dan sejahtera)   

                        Unnamed: 4  
0                          Catatan  
1         KRT: Kepala Rumah Tangga  
2    KKS: Kartu Keluarga Sejahtera  
3    PKH: Program Keluarga Harapan  
4   BPNT: Bantuan Pangan Non Tunai  
5                              NaN  
6                              NaN  
7                              NaN  
8                              NaN  
9                              NaN  
10                             NaN  
11                             NaN  
12                             NaN  
13                             NaN  
14                             NaN  
15                             NaN  
16                             NaN  
17                             NaN  
18                             NaN  
19                             NaN  
20                             NaN  
21                             NaN

# disini untuk memudahkan maka saya mengubahnya menjadi binary 
# dimana pada 2 saya ubah menjadi 0 dan satu tetap menjadi 1

In [7]:
df = df.replace({2: 0, 1: 1})
df

X1_klas  X2_JK_KRT  X3_kelum_KRT  X4_ijz_KRT  X5_bekerja_KRT  \
0           0          1             0           1               1   
1           0          1             0           1               0   
2           0          1             0           1               1   
3           1          1             0           1               1   
4           1          0             0           0               0   
...       ...        ...           ...         ...             ...   
3264        0          1             0           1               1   
3265        0          1             0           0               1   
3266        0          1             0           0               1   
3267        1          1             0           0               1   
3268        0          1             0           0               1   

      X6_inet_KRT  X7_milik_jamkes_KRT  X8_kepemilikan_rumah  \
0               1                    1                     1   
1               0                    0                     1   
2               1                    1                     1   
3               0                    1                     0   
4               0                    1                     1   
...           ...                  ...                   ...   
3264            0                    1                     1   
3265            0                    0                     1   
3266            0                    1                     1   
3267            0                    0                     0   
3268            0                    0                     1   

      X9_luas_kapita_cat  X10_atap  ...  X14_air  X15_bbm  \
0                      0         1  ...        0        1   
1                      0         1  ...        0        1   
2                      0         1  ...        1        1   
3                      0         1  ...        1        0   
4                      0         1  ...        1        1   
...                  ...       ...  ...      ...      ...   
3264                   0         1  ...        1        1   
3265                   0         1  ...        1        1   
3266                   0         1  ...        1        1   
3267                   1         1  ...        0        1   
3268                   0         1  ...        1        1   

      X16_layanan_keuangan  X17_motor  X18_tanah  X19_kks  X20_pkh  X21_bpnt  \
0                        1          1          1        0        0         0   
1                        0          0          1        0        0         0   
2                        1          1          0        0        0         0   
3                        0          1          0        0        0         0   
4                        0          1          1        0        0         0   
...                    ...        ...        ...      ...      ...       ...   
3264                     0          1          1        0        0         0   
3265                     0          1          1        0        0         0   
3266                     0          1          1        0        0         0   
3267                     0          1          0        0        0         0   
3268                     1          0          0        0        0         1   

      X22_bantuan_pemda  X23_kelompok_pengeluaran  
0                     0                         0  
1                     0                         1  
2                     0                         0  
3                     0                         0  
4                     0                         0  
...                 ...                       ...  
3264                  0                         0  
3265                  0                         0  
3266                  0                         1  
3267                  0                         1  
3268                  0                         1  

[3269 rows x 23 columns]

# setelah saya mengubahnya maka saya akan memilih beberapa topik yang akan diambil yaitu 

# Topik Yang akan di ambil 
#### Pengaruh   Jenis Kelamin,Pendidikan KRT,Pekerjaan KRT dan status yang memiliki kendaraan  motor terhadap Akses layanan keuangan)

\[
\begin{array}{|c|l|c|}
\hline
\textbf{Nama Variabel} & \textbf{Kategori} & \textbf{Kode} \\
\hline
\text{Akses layanan keuangan(Y)} & \text{Punya Akses} & 1 \\
& \text{tidak punya akses} & 0 \\
\hline
\text{Jenis Kelamin (X1)} & \text{Laki-laki} & 1 \\
& \text{Perempuan} & 0 \\
\hline
\text{Pendidikan KRT (X2)} & \text{SD ke bawah} & 1 \\
& \text{SMP ke atas} & 0 \\
\hline
\text{Pekerjaan KRT (X3)} & \text{bekerja} & 1 \\
& \text{tidak bekerja} & 0 \\
\hline
\text{memiliki kendaraan  motor(X4)} & \text{Punya} & 1 \\
& \text{Tidak Punya} & 0 \\
\hline
\end{array}
\]


# Melihat statsistik deskriptif dari data yang akan di teliti 

In [14]:
# pada variabel indenpendent X  adalah DataFrame yang telah dipilih
# pada variabel dependent Y
X = df[["X2_JK_KRT", "X4_ijz_KRT", "X5_bekerja_KRT", "X17_motor"]]
# Menghitung jumlah angka 0 dan 1 di setiap kolom
Y = df[["X16_layanan_keuangan"]]
jumlah_0_x = X.apply(lambda col: col.value_counts().get(0, 0))
jumlah_1_x = X.apply(lambda col: col.value_counts().get(1, 0))
jumlah_0_y = Y.apply(lambda col: col.value_counts().get(0, 0))
jumlah_1_y = Y.apply(lambda col: col.value_counts().get(1, 0))
# Menampilkan hasil
print("Jumlah angka 0 per kolom variabel indenpendet X:")
print(jumlah_0_x)

print("\nJumlah angka 1 per kolom  kolom variabel indenpendet X :")
print(jumlah_1_x)

print("\nJumlah angka 0 per kolom variabel denpendet Y:")
print(jumlah_0_y)

print("\nJumlah angka 1 per kolom  kolom variabel denpendet Y :")
print(jumlah_1_y)


Jumlah angka 0 per kolom variabel indenpendet X:
X2_JK_KRT          282
X4_ijz_KRT        1904
X5_bekerja_KRT     267
X17_motor          270
dtype: int64

Jumlah angka 1 per kolom  kolom variabel indenpendet X :
X2_JK_KRT         2987
X4_ijz_KRT        1365
X5_bekerja_KRT    3002
X17_motor         2999
dtype: int64

Jumlah angka 0 per kolom variabel denpendet Y:
X16_layanan_keuangan    2503
dtype: int64

Jumlah angka 1 per kolom  kolom variabel denpendet Y :
X16_layanan_keuangan    766
dtype: int64


Dapat di lihat dari hasil pada tabel berikut setelah di hitung dari perhitungan di atas 

\[
\begin{array}{|c|l|c|}
\hline
\textbf{Nama Variabel} & \textbf{Kategori} & \textbf{Kode} & jumlah \\
\hline
\text{Akses layanan keuangan(Y)} & \text{Punya Akses} & 1 & 2503 \\
& \text{tidak punya akses} & 0 & 766  \\
\hline
\text{Jenis Kelamin (X1)} & \text{Laki-laki} & 1 & 2987 \\
& \text{Perempuan} & 0 & 282 \\
\hline
\text{Pendidikan KRT (X2)} & \text{SD ke bawah} & 1 & 1365 \\
& \text{SMP ke atas} & 0 &  1904 \\
\hline
\text{Pekerjaan KRT (X3)} & \text{bekerja} & 1 & 3002 \\
& \text{tidak bekerja} & 0 & 267 \\
\hline
\text{memiliki kendaraan  motor(X4)} & \text{Punya} & 1 &  2999 \\
& \text{Tidak Punya} & 0 & 270\\
\hline
\end{array}
\]


## Model regresi logistik biner yang akan digunakan  sebagai berikut 
$$
\text{logit}(p)  =  \log\left(\frac{p}{1-p}\right) =\beta_0 + \beta_1 X_1 + \beta_2 X_2 + \beta_3 X_3  + \beta_4 X_4 
$$



In [8]:
import pandas as pd
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Pastikan df adalah DataFrame yang sudah dipersiapkan sebelumnya
X = df[["X2_JK_KRT", "X4_ijz_KRT", "X5_bekerja_KRT", "X17_motor"]]  # Fitur
y = df["X16_layanan_keuangan"]  # Target

# Membagi data menjadi data pelatihan dan pengujian
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=1, random_state=3692)

# Melakukan standardisasi fitur
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)  # Mengubah menjadi numpy array 2D
X_test = scaler.transform(X_test)  # Mengubah menjadi numpy array 2D

# Menambahkan konstanta (intercept) pada fitur
X_train_with_intercept = sm.add_constant(X_train)

# Membuat dan melatih model regresi logistik menggunakan statsmodels
logit_model = sm.MNLogit(y_train, X_train_with_intercept)  # MNLogit untuk regresi logistik multinomial
result = logit_model.fit()

# Menampilkan hasil estimasi parameter (koefisien dan p-value)
print(result.summary())


Optimization terminated successfully.
         Current function value: 0.534794
         Iterations 6
                           MNLogit Regression Results                           
Dep. Variable:     X16_layanan_keuangan   No. Observations:                 3268
Model:                          MNLogit   Df Residuals:                     3263
Method:                             MLE   Df Model:                            4
Date:                  Thu, 05 Dec 2024   Pseudo R-squ.:                 0.01788
Time:                          15:52:37   Log-Likelihood:                -1747.7
converged:                         True   LL-Null:                       -1779.5
Covariance Type:              nonrobust   LLR p-value:                 4.958e-13
X16_layanan_keuangan=1       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                     -1.2230      0.043    -28.454      0.000  

Berdasarkan hasil pengelolan dihasilkan model regresi Logistik Biner untuk mengetahui pengaruh akses layananan keuangan sebagai berikut 

$$
\text{logit}(p)  = -1.2230 + 0.1515 X_1 -0.1407  X_2 +   0.1820 X_3  + 0.1650   X_4 
$$
##### Dimana 
$$
X_{1}: \text{Jenis Kelamin Kepala Rumah Tangga} \\
X_{2}: \text{Pendidikan Kepala Rumah Tangga} \\
X_{3}: \text{Status Pekerjaan Kepala Rumah Tangga} \\
X_{4}: \text{Kepemilikan kendaraan bermotor}
$$

 


In [16]:
print(result.summary())

                           MNLogit Regression Results                           
Dep. Variable:     X16_layanan_keuangan   No. Observations:                 3268
Model:                          MNLogit   Df Residuals:                     3263
Method:                             MLE   Df Model:                            4
Date:                  Thu, 05 Dec 2024   Pseudo R-squ.:                 0.01788
Time:                          16:19:25   Log-Likelihood:                -1747.7
converged:                         True   LL-Null:                       -1779.5
Covariance Type:              nonrobust   LLR p-value:                 4.958e-13
X16_layanan_keuangan=1       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                     -1.2230      0.043    -28.454      0.000      -1.307      -1.139
x1                         0.1515      0.055      2.732      0.006       0.043 

In [9]:
## mencari  Odd ratio dari koefisien masing masing model dan interceptnya yaitu 
oddk = math.exp( -1.2230)
odd1 = math.exp( 0.1515 )
odd2 = math.exp(-0.1407)
odd3 = math.exp( 0.1820)
odd4 = math.exp( 0.1650)
print ("nilai dari odd ratio konstanta atau intercept yaitu ",oddk)
print ("nilai dari odd ratio dari x1 yaitu ",odd1)
print ("nilai dari odd ratio dari x2 yaitu ",odd2)
print ("nilai dari odd ratio dari x3 yaitu ",odd3)
print ("nilai dari odd ratio dari x4 yaitu ",odd4)

nilai dari odd ratio konstanta atau intercept yaitu  0.29434580363145335
nilai dari odd ratio dari x1 yaitu  1.1635783018096755
nilai dari odd ratio dari x2 yaitu  0.8687498975771047
nilai dari odd ratio dari x3 yaitu  1.1996141938798683
nilai dari odd ratio dari x4 yaitu  1.1793931187113906


#### Interpretasi Hasil Estimasi:
- Hasil estimasi memperlihatkan bahwa Jenis kelamin Kepala Rumah tangga (X1) berpengaruh signifikan terhadap akses layanan keuangan dimana p-value lebih kecil dari 0.05,Oleh Karena itu  Kepala rumah tangga laki-laki memiliki kecenderungan 1.163 kali untuk memiliki akses layanan keuangan lebih tinggi di bandingkan kepala rumah tangga perempuan.Hal ini mungkin di sebabkan oleh  dampak norma sosial dapat membatasi peran perempuan dalam pengambilan keputusan keuangan dan Perempuan sering kali memiliki sumber daya ekonomi yang lebih terbatas, sehingga sulit untuk memenuhi untuk mengakses layanan keuangan. 

- Sementara itu Pendidikan kepala rumah tangga berpengaruh signifikan  terhadap akses layanan keuangan dimana p-value juga  lebih kecil dari 0.05,Sehingga  pendidikan sekolah dasar kebawah(SD) memiliki kecenderungan 0.868 kali  untuk memiliki akses layanan keuangan lebih tinggi dibandingkan  pendidikan SMP ke atas. Hal ini menunjukkan bahwa individu dengan pendidikan yang lebih rendah mungkin memiliki cara atau strategi tertentu dalam mengakses layanan keuangan, meskipun secara umum, pendidikan yang lebih tinggi diharapkan dapat meningkatkan akses tersebut.


- Selanjutnya variabel Status Pekerjaan kepala rumah tangga memberikan pengaruh signifikan terhadap akses layanan keuangan dimana p-value juga  lebih kecil dari 0.05, Oleh karena itu status pekerja yang telah bekerja memliki kecenderungan 1.19 kali untuk memiliki akses layanan keuangan lebih tinggi di bandingkan yang tidak bekerja hal ini mungkin terjadi karena individu  yang sudah bekerja cenderung memiliki stabilitas pendapatan yang lebih baik, dengan gaji bulanan tetap yang memudahkan mereka mengatur keuangan dan mengakses layanan keuangan seperti pinjaman, asuransi, atau investasi

- Terakhir variabel kepemilikan kendaraan bermotor berpengaruh signifikan terhadap akses layanan keuangan dimana p-value juga  lebih kecil dari 0.05,Oleh karena itu status kepemilikan kendaraan motor memliki kecenderungan 1.17 kali untuk ntuk memiliki akses layanan keuangan lebih tinggi dibandingkan tidak memiliki kendaraan motor hal ini mungkin terjadi karena kendaraan bermotor meningkatkan mobilitas individu untuk  memudahkan mereka untuk mengunjungi lembaga keuangan, seperti bank atau perusahaan pinjaman.


In [15]:
print(result.summary())

                           MNLogit Regression Results                           
Dep. Variable:     X16_layanan_keuangan   No. Observations:                 3268
Model:                          MNLogit   Df Residuals:                     3263
Method:                             MLE   Df Model:                            4
Date:                  Thu, 05 Dec 2024   Pseudo R-squ.:                 0.01788
Time:                          16:13:19   Log-Likelihood:                -1747.7
converged:                         True   LL-Null:                       -1779.5
Covariance Type:              nonrobust   LLR p-value:                 4.958e-13
X16_layanan_keuangan=1       coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                     -1.2230      0.043    -28.454      0.000      -1.307      -1.139
x1                         0.1515      0.055      2.732      0.006       0.043 

## Uji keseuaian model

Berdasarkan hasil dari model **Multinomial Logit (MNLogit)**, dapat disimpulkan sebagai berikut:

1. **Pseudo R-squared** yang rendah (0.01788) menunjukkan bahwa model tidak menjelaskan banyak variasi dalam data. Meskipun demikian, hal ini adalah hal yang umum terjadi pada model multinomial logit, karena model ini sering kali menghasilkan nilai Pseudo R-squared yang lebih rendah dibandingkan dengan regresi linier.

2. **LLR p-value** yang sangat kecil (4.958e-13) menunjukkan bahwa model dengan prediktor lebih kompleks memberikan hasil yang lebih baik dibandingkan model kosong (tanpa prediktor). Artinya, model ini secara statistik **signifikan** dan dapat dianggap cocok dengan data yang ada.

3. Secara keseluruhan, meskipun nilai **Pseudo R-squared** relatif rendah, uji kesesuaian model menunjukkan bahwa model **Multinomial Logit** model tetap signifikan dan memberikan penjelasan yang lebih baik dibandingkan model tanpa variabel.

4. **Covariance Type** nonrobust artinya jenis kovarians yang digunakan untuk estimasi adalah nonrobust, yang berarti bahwa model ini tidak mengoreksi varians estimasi jika terdapat heteroskedastisitas.

5. **converged**  True artinya bahwa model  berhasil menemukan solusi untuk parameter model


## Daftar Pustaka 

- https://github.com/DaniNegoita/Multinomial-Logistic-Regression-in-Python.git
- https://github.com/KamitaniLab/smlr.git
- Susanti, V., & Annisa, L. (2024). KARAKTERISTIK DAN DETERMINAN STATUS PENGANGGURAN PADA PENDUDUK USIA MUDA DI PROVINSI JAMBI TAHUN 2022. Media Edukasi Data Ilmiah dan Analisis (MEDIAN), 7(01), 15-33.

- Gujarati, D. N. (2021). Essentials of econometrics. Sage Publications.

## follow my github 
